python/document-detective/document_detective/main.py

```diff
@@ -0,0 +1,185 @@
+# to automatically retry indexing up to 5 times with exponential backoff.
+
+
+@DBOS.step(retries_allowed=True, max_attempts=5)
+def index_document(document_url: HttpUrl) -> int:
+    with TemporaryDirectory() as temp_dir:
+        temp_file_path = os.path.join(temp_dir, "file.pdf")
+        with open(temp_file_path, "wb") as temp_file:
+            with requests.get(document_url, stream=True) as r:
+                r.raise_for_status()
+                for page in r.iter_content(chunk_size=8192):
+                    temp_file.write(page)
+            temp_file.seek(0)
+            reader = PDFReader()
+            pages = reader.load_data(temp_file_path)
+    for page in pages:
+        index.insert(page)
+    return len(pages)
+
+
+# This is the endpoint for indexing. It starts the indexing workflow in the background
+# on a batch of documents.
+
+# To test it out, try this example cURL command to index Apple's SEC 10-K filings
+# for 2021, 2022, and 2023.
+
+"""
+curl -X POST "http://localhost:8000/index" \
+     -H "Content-Type: application/json" \
+     -d '{"urls": ["https://d18rn0p25nwr6d.cloudfront.net/CIK-0000320193/faab4555-c69b-438a-aaf7-e09305f87ca3.pdf", "https://d18rn0p25nwr6d.cloudfront.net/CIK-0000320193/b4266e40-1de6-4a34-9dfb-8632b8bd57e0.pdf", "https://d18rn0p25nwr6d.cloudfront.net/CIK-0000320193/42ede86f-6518-450f-bc88-60211bf39c6d.pdf"]}'
+"""
+
+
+class URLList(BaseModel):
+    urls: List[HttpUrl]
+
+
+@app.post("/index")
+async def index_endpoint(urls: URLList):
+    DBOS.start_workflow(indexing_workflow, urls.urls)
+
+
+# Now, let's chat!
+
+# Each time we get a chat message, we call this workflow with three steps:
+# 1. Store the incoming chat message in Postgres.
+# 2. Query LlamaIndex to respond to the message using RAG.
+# 3. Store the response in Postgres.
+
+
+class ChatSchema(BaseModel):
+    message: str
+
+
+@app.post("/chat")
+@DBOS.workflow()
+def chat_workflow(chat: ChatSchema):
+    insert_chat(chat.message, True)
+    response = query_model(chat.message)
+    insert_chat(response, False)
+    return {"content": response, "isUser": True}
+
+
+@DBOS.transaction()
+def insert_chat(content: str, is_user: bool):
+    DBOS.sql_session.execute(
+        chat_history.insert().values(content=content, is_user=is_user)
+    )
+
+
+@DBOS.step()
+def query_model(message: str) -> str:
+    return str(chat_engine.chat(message))
+
+
+# Let's also write a history endpoint that retrieves all past chats
+# from the database.
+
+# This function is called when we open up the chatbot so it
+# can display your chat history.
+
+
+@app.get("/history")
+def history_endpoint():
+    return get_chats()
+
+
+@DBOS.transaction()
+def get_chats():
+    stmt = chat_history.select().order_by(chat_history.c.created_at.asc())
+    result = DBOS.sql_session.execute(stmt)
+    return [{"content": row.content, "isUser": row.is_user} for row in result]
```

Comment:
You might want to add the same user filtering than Qian's chatbot.

Reply:
Keeping the chat part small to focus on ingestion.